examples/text/plot_hashing_vs_dict_vectorizer.py

```diff
@@ -3,109 +3,293 @@
-print("    The default number of features is 2**18.")
-print()
+token_freqs("That is one example, but this is another one")
 
-try:
-    n_features = int(sys.argv[1])
-except IndexError:
-    n_features = 2**18
-except ValueError:
-    print("not a valid number of features: %r" % sys.argv[1])
-    sys.exit(1)
+# %%
+# DictVectorizer
+# --------------
+#
+# First we benchmark the `DictVectorizer`, then we compare it to `FeatureHasher`
+# as both of them receive dictionaries as input.
 
+from time import time
+from sklearn.feature_extraction import DictVectorizer
 
-print("Loading 20 newsgroups training data")
-raw_data, _ = fetch_20newsgroups(subset="train", categories=categories, return_X_y=True)
-data_size_mb = sum(len(s.encode("utf-8")) for s in raw_data) / 1e6
-print("%d documents - %0.3fMB" % (len(raw_data), data_size_mb))
-print()
+dict_count_vectorizers = defaultdict(list)
 
-print("DictVectorizer")
 t0 = time()
 vectorizer = DictVectorizer()
 vectorizer.fit_transform(token_freqs(d) for d in raw_data)
 duration = time() - t0
-print("done in %fs at %0.3fMB/s" % (duration, data_size_mb / duration))
-print("Found %d unique terms" % len(vectorizer.get_feature_names_out()))
-print()
+dict_count_vectorizers["vectorizer"].append(vectorizer.__class__.__name__)
+dict_count_vectorizers["speed"].append(data_size_mb / duration)
+print(f"done in {duration:.3f}s at {data_size_mb / duration:.3f}MB/s")
+print(f"Found {len(vectorizer.get_feature_names_out())} unique terms")
+
+# %%
+# FeatureHasher
+# -------------
+#
+# Dictionaries take up a large amount of storage space and grow in size as the
+# training set grows. Instead of growing the vectors along with a dictionary,
+# feature hashing builds a vector of pre-defined length by applying a hash
+# function `h` to the features (e.g., tokens), then using the hash values
+# directly as feature indices and updating the resulting vector at those
+# indices. Hashing functions tend to map distinct values to the same hash code
+# (hash collisions). As a result, it is impossible to determine what object
+# generated any particular hash code.
+#
+# Because of the above it is impossible to recover the features and the best
+# approach to estimate the number of unique terms in the original dictionary is
+# to count non-vanishing entries of the encoded features. For such purpose we
+# define the following function:
+
+import numpy as np
+
+
+def n_nonzero_columns(X):
+    """Number of columns with at least one non-zero value in a CSR matrix.
+
+    This is useful to count the number of features columns that are effectively
+    active when using the FeatureHasher.
+    """
+    return len(np.unique(X.nonzero()[1]))
+
+
+# %%
+# The default number of features for the `FeatureHasher` is 2**20. Here we set
+# `n_features = 2**18` to illustrate hash collisions. Notice that a large number
+# of features causes larger coefficient dimensions and then requires more
+# training time.
```

Comment:
Not sure about training time: some estimators (e.g. linear models) will naturally skip zero columns if they are encoded in a sparse matrix. But it will require more memory usage to store the coefficient of the model even if a majority of them is inactive.